### ISSUE TITLE
Add function calling and MCP support

### ISSUE THREAD

**Issue by rlouf** [outlines-dev/outlines#1626]
A v0 would be allowing users to instantiate the logits processor with the tool call tokens and pass it to the generator. Note that this would only work with models that use simple tokens for tool calling (Llama has a more complex pattern that we can handle later).

```python
model = ...
logits_processor = FunctionCallingLogitsProcessor("<tool_call>", "</tool_call>",  Union[User, Customer])
generator = Generator(model, processor=logits_processor)
```

The implementation is simple:
1. While `<tool_call>` has not been observed, `__call__` is a pass-through.
2. After `<tool_call>` has been observed, mask the logits to follow a given JSON Schema. 
3. Once the JSON generation has completed, force the generation of `</tool_call>`
4. Then let the model generate freely, `__call__` is a pass-through.

If there are several possible function calls, the type for (2) would be a `Union` of the different JSON Schemas.

**Comment by Youssefares** [comment]
I am going to take a stab at this. Might post a few snippets of implementation approach here first if PR'd be non trivial.

**Comment by Youssefares** [comment]
@rlouf #1910 is green now & ready for review, it implements the V0 as originally spec'ed out. Keen to hear your thoughts.

**Comment by rehan243** [comment]
Oh interesting, I've worked on something similar for a logits processor, and one thing that stood out here is how you're handling the JSON Schema enforcement. If you're masking logits after `<tool_call>` based on a schema, are you planning to dynamically update the mask based on partial JSON generation? Like, if the model generates an opening `{`, are you restricting it to valid keys from the schema next? That part can get messy pretty fast without a stateful tracker for the schema constraints.

Also, for forcing `</tool_call>` at the end, are you thinking of injecting a bias for that token in the logits? Or hard-masking everything else? I've seen token biasing work well in GPT-style models, but not sure how robust it'd be depending on the model's decoding strategy.

Curious, what models are you targeting first? Are you starting with Llama-v2 or something simpler like GPT-based ones?
